R/chat.R

```diff
@@ -731,6 +743,33 @@ Chat <- R6::R6Class(
 
     has_system_prompt = function() {
       length(private$.turns) > 0 && is_system_turn(private$.turns[[1]])
+    },
+
+    complete_dangling_tool_requests = function() {
+      if (length(private$.turns) == 0) {
+        return(invisible())
+      }
+
+      last_turn <- private$.turns[[length(private$.turns)]]
+      if (last_turn@role != "assistant") {
+        return(invisible())
+      }
+
+      tool_requests <- keep(last_turn@contents, is_tool_request)
+      if (length(tool_requests) == 0) {
+        return(invisible())
+      }
+
+      tool_results <- lapply(tool_requests, function(req) {
+        ContentToolResult(
+          error = "Chat ended before the tool could be invoked.",
+          request = req
+        )
+      })
+      tool_turn <- tool_results_as_turn(tool_results)
+      private$.turns[[length(private$.turns) + 1]] <- tool_turn
+
+      invisible(tool_results)
```

Comment:
```suggestion
      invisible()
```

IMO this better signals that you're calling the function primarily for its side-effects.